### PR TITLE
RsETH on Linea

### PR DESCRIPTION
**Pull request by jalextowle**
# Description

This PR adds support for Kelp DAO's rsETH yield source on Linea. This was a pretty standard LRT yield source, but there are a few things for reviewers to think about:

- Is this named appropriately? The vault shares token is actually called `wrsETH`, but it is redeemable 1:1 with rsETH. rsETH seems like it has more name recognition, but it's worth thinking about. We could also call this `WrsETHHyperdrive`.
- What do we think about supporting `asBase = true` when the fees are zero? We could just rip this logic out if we wanted to.

### CODE REVIEW

File: contracts/src/instances/rseth-linea/RsETHLineaHyperdrive.sol [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+/// SSSSSSSS                                                                SSSSSSSS
+/// SSSSSSSS                                                                SSSSSSSS
+///
+/// @author DELV
+/// @title RsETHLineaHyperdrive
+/// @notice A Hyperdrive instance that uses a RsETHLinea vault as the yield source.
+/// @dev This instance supports the Renzo protocol on Linea. The vault shares
+///      token is the non-rebasing LRT token wrsETH. There are a few special
+///      things about this integration:
+///
```
Comment (mcclurejt) [inline contracts/src/instances/rseth-linea/RsETHLineaHyperdrive.sol:56]: is this space intentional?

File: contracts/src/instances/rseth-linea/RsETHLineaHyperdrive.sol [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+///      - The base token address is the ETH constant.
+///      - The vault shares token address is the wrsETH address.
+///      - The vault share price is provided by an oracle.
+///      - Interest accrues sporadically when the oracle is updated.
+///      - Base deposits are supported as long as deposit fees are turned off.
+///      - Base withdrawals aren't supported since there isn't an instantaneous
+///        way to withdraw from wrsETH.
+///      - The minimum share reserves and minimum transaction amount are both
+///        1e15.
```
Comment (mcclurejt) [inline contracts/src/instances/rseth-linea/RsETHLineaHyperdrive.sol:65]: ❤️

File: contracts/src/interfaces/IRsETHLinea.sol [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.22;
+
+import { IERC4626 } from "./IERC4626.sol";
+
+// ****************************************************************************
+// FIXME: Fill out the interface as needed.  ERC4626 extended as an example.
```
Comment (mcclurejt) [inline contracts/src/interfaces/IRsETHLinea.sol:7]: fixme

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+import { RsETHLineaTarget4Deployer } from "../../../contracts/src/deployers/rseth-linea/RsETHLineaTarget4Deployer.sol";
+import { HyperdriveFactory } from "../../../contracts/src/factory/HyperdriveFactory.sol";
+import { RsETHLineaConversions } from "../../../contracts/src/instances/rseth-linea/RsETHLineaConversions.sol";
+import { IERC20 } from "../../../contracts/src/interfaces/IERC20.sol";
+import { IHyperdrive } from "../../../contracts/src/interfaces/IHyperdrive.sol";
+import { IRsETHLineaHyperdrive } from "../../../contracts/src/interfaces/IRsETHLineaHyperdrive.sol";
+import { IRSETHPoolV2 } from "../../../contracts/src/interfaces/IRSETHPoolV2.sol";
+import { AssetId } from "../../../contracts/src/libraries/AssetId.sol";
+import { ETH } from "../../../contracts/src/libraries/Constants.sol";
+import { FixedPointMath, ONE } from "../../../contracts/src/libraries/FixedPointMath.sol";
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:20]:
```suggestion
import { FixedPointMath } from "../../../contracts/src/libraries/FixedPointMath.sol";
```

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+
+import { stdStorage, StdStorage } from "forge-std/Test.sol";
+import { RsETHLineaHyperdriveCoreDeployer } from "../../../contracts/src/deployers/rseth-linea/RsETHLineaHyperdriveCoreDeployer.sol";
+import { RsETHLineaHyperdriveDeployerCoordinator } from "../../../contracts/src/deployers/rseth-linea/RsETHLineaHyperdriveDeployerCoordinator.sol";
+import { RsETHLineaTarget0Deployer } from "../../../contracts/src/deployers/rseth-linea/RsETHLineaTarget0Deployer.sol";
+import { RsETHLineaTarget1Deployer } from "../../../contracts/src/deployers/rseth-linea/RsETHLineaTarget1Deployer.sol";
+import { RsETHLineaTarget2Deployer } from "../../../contracts/src/deployers/rseth-linea/RsETHLineaTarget2Deployer.sol";
+import { RsETHLineaTarget3Deployer } from "../../../contracts/src/deployers/rseth-linea/RsETHLineaTarget3Deployer.sol";
+import { RsETHLineaTarget4Deployer } from "../../../contracts/src/deployers/rseth-linea/RsETHLineaTarget4Deployer.sol";
+import { HyperdriveFactory } from "../../../contracts/src/factory/HyperdriveFactory.sol";
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:12]:
```suggestion
```

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+import { RsETHLineaConversions } from "../../../contracts/src/instances/rseth-linea/RsETHLineaConversions.sol";
+import { IERC20 } from "../../../contracts/src/interfaces/IERC20.sol";
+import { IHyperdrive } from "../../../contracts/src/interfaces/IHyperdrive.sol";
+import { IRsETHLineaHyperdrive } from "../../../contracts/src/interfaces/IRsETHLineaHyperdrive.sol";
+import { IRSETHPoolV2 } from "../../../contracts/src/interfaces/IRSETHPoolV2.sol";
+import { AssetId } from "../../../contracts/src/libraries/AssetId.sol";
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:18]:
```suggestion
```

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+import { ETH } from "../../../contracts/src/libraries/Constants.sol";
+import { FixedPointMath, ONE } from "../../../contracts/src/libraries/FixedPointMath.sol";
+import { HyperdriveMath } from "../../../contracts/src/libraries/HyperdriveMath.sol";
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:21]:
```suggestion
```

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+import { LPMath } from "../../../contracts/src/libraries/LPMath.sol";
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:22]:
```suggestion
```

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+import { ERC20ForwarderFactory } from "../../../contracts/src/token/ERC20ForwarderFactory.sol";
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:23]:
```suggestion
```

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+import { ERC20Mintable } from "../../../contracts/test/ERC20Mintable.sol";
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:24]:
```suggestion
```

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+import { InstanceTest } from "../../utils/InstanceTest.sol";
+import { HyperdriveUtils } from "../../utils/HyperdriveUtils.sol";
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:26]:
```suggestion
import { HyperdriveUtils } from "../../utils/HyperdriveUtils.sol";
import { InstanceTest } from "../../utils/InstanceTest.sol";
```

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+    /// @dev Gets the extra data used to deploy RsETHLineaHyperdrive instances.
+    ///      This is empty since the instance gets all of it's information from
+    ///      the deployer coordinators and deployers.
+    /// @return The empty extra data.
+    function getExtraData() internal pure override returns (bytes memory) {
+        return new bytes(0);
+    }
+
+    /// @dev Converts base amount to the equivalent about in shares.
+    function convertToShares(
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:123]: missing @param @return

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+        uint256 baseAmount
+    ) internal view override returns (uint256) {
+        return RsETHLineaConversions.convertToShares(RSETH_POOL, baseAmount);
+    }
+
+    /// @dev Converts share amount to the equivalent amount in base.
+    function convertToBase(
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:130]: missing @param @return

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 14]
```diff
@@ -0,0 +1,351 @@
...
+        uint256 shareAmount
+    ) internal view override returns (uint256) {
+        return RsETHLineaConversions.convertToBase(RSETH_POOL, shareAmount);
+    }
+
+    /// @dev Deploys the rsETH Linea deployer coordinator contract.
+    /// @param _factory The address of the Hyperdrive factory.
+    function deployCoordinator(
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:138]: missing @return

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+            );
+    }
+
+    /// @dev Fetches the total supply of the base and share tokens.
+    function getSupply() internal view override returns (uint256, uint256) {
+        return (address(RSETH_POOL).balance, WRSETH.totalSupply());
+    }
+
+    /// @dev Fetches the token balance information of an account.
+    function getTokenBalances(
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:164]: missing @param @return

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+        address account
+    ) internal view override returns (uint256, uint256) {
+        return (account.balance, WRSETH.balanceOf(account));
+    }
+
+    /// @dev Verifies that deposit accounting is correct when opening positions.
+    function verifyDeposit(
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:171]: missing @param

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+                traderSharesAfter,
+                traderBalancesBefore.sharesBalance -
+                    hyperdrive.convertToShares(amountPaid),
+                2
+            );
+        }
+    }
+
+    /// @dev Verifies that withdrawal accounting is correct when closing positions.
+    function verifyWithdrawal(
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:266]: missing @param

File: test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol [hunk 18]
```diff
@@ -0,0 +1,351 @@
...
+            address(IRsETHLineaHyperdrive(address(hyperdrive)).rsETHPool()),
+            address(RSETH_POOL)
+        );
+        (, uint256 totalShares) = getTokenBalances(address(hyperdrive));
+        assertEq(hyperdrive.totalShares(), totalShares);
+    }
+
+    /// Helpers ///
+
+    function advanceTime(
```
Comment (mcclurejt) [inline test/instances/rseth-linea/RsETHLineaHyperdrive.t.sol:331]: missing natspec

File: contracts/src/deployers/rseth-linea/RsETHLineaHyperdriveDeployerCoordinator.sol [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+        // Perform the default checks.
+        super._checkPoolConfig(_deployConfig, _extraData);
+
+        // Ensure that the vault shares token address is properly configured.
+        if (address(_deployConfig.vaultSharesToken) != rsETHPool.wrsETH()) {
+            revert IHyperdriveDeployerCoordinator.InvalidVaultSharesToken();
+        }
+
+        // Ensure that the base token address is properly configured.
+        if (address(_deployConfig.baseToken) != address(ETH)) {
```
Comment (mcclurejt) [inline contracts/src/deployers/rseth-linea/RsETHLineaHyperdriveDeployerCoordinator.sol:161]: do we need this? if we're checking for the eth address shoudln't we also check for the zero address?

**Review by mcclurejt — APPROVED** [review]
Looks great, comments are all nits

File: contracts/src/instances/rseth-linea/RsETHLineaHyperdrive.sol (hunk 1 above)
Comment (jalextowle) [reply contracts/src/instances/rseth-linea/RsETHLineaHyperdrive.sol:56]: Yes, it was. Just a line break between the comment and the list

File: contracts/src/deployers/rseth-linea/RsETHLineaHyperdriveDeployerCoordinator.sol (hunk 19 above)
Comment (mcclurejt) [reply contracts/src/deployers/rseth-linea/RsETHLineaHyperdriveDeployerCoordinator.sol:161]: nvm I can't read

**Pull request merged**